Update to have only one poll call OKTA-423330

anatoliiboiarshchykov-okta opened the pull request:
## Description:
- update remediation poll from setInterval to setTimeout
- update v2 AppState to trigger 'change:dynamicRefreshInterval' even if interval is not changed
- update NumberChallengePushView, ChallengePushView, EnrollPollOktaVerifyView to be extended from BaseFormWithPolling
- add test case for new polling logic in DeviceChallengePollView
- Update AppState unit test

SIW bacon [link](https://bacon-go.aue1e.saasure.net/commits?artifact=okta-signin-widget&sha=9e9b47c9674c7a1b53f842d491aea0c6ca5889e6&page=1&pageSize=6)
Core bacon [link](https://bacon-go.aue1e.saasure.net/commits?artifact=monolith&sha=ff56129adf983bf2ab45c9454b8b7c18377c61ab&page=1&pageSize=6)


## PR Checklist

- [ ] Have you verified the basic functionality for this change?

- [ ] Added unit tests?

- [ ] Added e2e tests

- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Screenshot/Video:
- [before fix](https://okta.box.com/s/k8j42wa5anibalj0b5b83v1cw3ebc3my)
- [after fix](https://okta.box.com/s/e7vjbk955zr5h1f39xnd3k3t7wbobliu)

### Reviewers:
@haishengwu-okta 
@pradeepdewda-okta 
@YangChen-okta 

### Issue:

- [OKTA-423330](https://oktainc.atlassian.net/browse/OKTA-423330)

pradeepdewda-okta reviewed (commented):
Review comments in slack..

https://okta.slack.com/archives/CAZH2MWEL/p1637171901273100?thread_ts=1637075738.249000&cid=CAZH2MWEL

pradeepdewda-okta commented on src/v2/view-builder/views/device/DeviceChallengePollView.js:
```diff
@@ -34,6 +34,7 @@ const Body = BaseFormWithPolling.extend(
       BaseFormWithPolling.prototype.initialize.apply(this, arguments);
       this.listenTo(this.model, 'error', this.onPollingFail);
       doChallenge(this, IDENTIFIER_FLOW);
+      this.options.settings.set('isParallelPollCallsAllowed', true);
```
what does this mean for poll requests ? `isParallelPollCallsAllowed`

anatoliiboiarshchykov-okta replied:
This code was removed

pradeepdewda-okta commented:
Test this in okta-core first to see, it doesn't fail any selenium.

haishengwu-okta commented on src/v2/models/AppState.js:
```diff
@@ -185,6 +185,8 @@ export default Model.extend({
 
     if (identicalResponse && !isSameRefreshInterval) {
       this.set('dynamicRefreshInterval', this.getRefreshInterval(transformedResponse));
+    } else {
+      this.trigger('change:dynamicRefreshInterval');
```
I think this shall be good enough?

```
if (identicalResponse) {
  this.set('dynamicRefreshInterval', this.getRefreshInterval(transformedResponse));
}

anatoliiboiarshchykov-okta replied:
Updated

haishengwu-okta replied:
I meant instead of creating if/else we could just do a single `if` block.

anatoliiboiarshchykov-okta replied:
Updated

haishengwu-okta reviewed (commented):
Add testcafe tests?

pradeepdewda-okta commented:
Before adding any code further.. test this in okta-core. It will be a breaking change.

pradeepdewda-okta commented on src/v2/models/AppState.js:
```diff
@@ -185,6 +185,8 @@ export default Model.extend({
     if (identicalResponse && !isSameRefreshInterval) {
       this.set('dynamicRefreshInterval', this.getRefreshInterval(transformedResponse));
+    } else if (identicalResponse) {
+      this.trigger('change:dynamicRefreshInterval');
```
so in case of identical response we always trigger `re-rendering` ? this might create a flicker type effect in `beacon` image..

pradeepdewda-okta replied:
also you don't need else if.. we are triggering poll every time if response is identical

anatoliiboiarshchykov-okta replied:
Updated

pradeepdewda-okta commented on src/v2/view-builder/views/shared/polling.js:
```diff
@@ -41,7 +41,7 @@ export default {
     const pollInterval = this.dynamicPollingInterval || this.fixedPollingInterval;
     if (_.isNumber(pollInterval)) {
-      this.polling = setInterval(() => {
+      this.polling = setTimeout(() => {
```
What about `setInterval` at line 30 ?

anatoliiboiarshchykov-okta replied:
Update

anatoliiboiarshchykov-okta commented:
@haishengwu-okta @pradeepdewda-okta 
I have made some updates:

- Update case when triggering  'dynamicRefreshInterval' change event(one if =)  link [one](https://github.com/okta/okta-signin-widget/pull/2304#discussion_r762024790) link [two](https://github.com/okta/okta-signin-widget/pull/2304#discussion_r763340770))
- Add a test cafe 
- added links to SIW and core Bacon in description
- Updated another setInterval => setTimeout(more [details](https://github.com/okta/okta-signin-widget/pull/2304#discussion_r764451704))

haishengwu-okta commented:
Fixes looks good to me.
As you changed 2 setInterval, I wonder there would be more changes to test?

pradeepdewda-okta commented on test/testcafe/spec/DeviceChallengePollView_spec.js:
```diff
@@ -317,6 +360,23 @@ test
+    // Updating /poll response to take 5 sec to response.
+    // Then counting the number of calls that should be done in time interval. Default Timeout for /poll is 2 sec.
+    // Expecting to get only 2 calls(first at 2nd sec, second at 9th(5 sec response + 2 sec timeout) second).
+    const tetsTimeout = 10_000;
```
what is `10_000` is it valid ?

anatoliiboiarshchykov-okta replied:
It needed to verify that at least 2 /poll calls with timeout were made. this allows verifying that the new poll call would not be started before the previous finished. 
The first call is made with a timeout of 2 sec, the second call will be made at the 9th second(2sec timeout of first calls + 2 sec timeout of second call + 5 seconds of first call response). And I decided to add an additional 1 sec to make sure this test is not flaky.
Before updating the logic we would get 4-5 calls(one call per 2 sec). 

Please advise, from your perspective, Is there a more efficient way to check it ?

anatoliiboiarshchykov-okta commented:
> Fixes looks good to me. As you changed 2 setInterval, I wonder there would be more changes to test?

@haishengwu-okta 
I can add an additional test case for this if it is strongly preferred. 
From my point of view, I do not think we need it. We already have test cases that cover polling logic in SIW & core. If they are not failing this is a good sign that this logic works. 
The test case that I have added is more related to the check bug that was mentioned in the ticket description. However, line 30 setInterval was changed not because of an issue, but because of new polling logic, and if all tests passed we can say that it works correctly.

Please advise, what will be your preference regarding this?

pradeepdewda-okta commented:
> > Fixes looks good to me. As you changed 2 setInterval, I wonder there would be more changes to test?
> 
> @haishengwu-okta I can add an additional test case for this if it is strongly preferred. From my point of view, I do not think we need it. We already have test cases that cover polling logic in SIW & core. If they are not failing this is a good sign that this logic works. The test case that I have added is more related to the check bug that was mentioned in the ticket description. However, line 30 setInterval was changed not because of an issue, but because of new polling logic, and if all tests passed we can say that it works correctly.
> 
> Please advise, what will be your preference regarding this?

Still good to add a test that checks for only one polling request because it is `setTimeout` now instead of `setInterval`..

haishengwu-okta commented:
@anatoliiboiarshchykov-okta aha, yeah, you're right. 
we'd probably need similar test for the other setTimeout to make sure it's only polling once if there are any errors.

anatoliiboiarshchykov-okta commented:
@haishengwu-okta  @pradeepdewda-okta  
Next changes were made:

- Added unit tests for polling
- added test case when a poll call fails
- added 2 demo videos

Please advise is there anything else that needs to be added?

pradeepdewda-okta commented on test/unit/spec/v2/view-builder/views/shared/polling_spec.js:
```diff
@@ -0,0 +1,77 @@
+import polling from '../../../../../../../src/v2/view-builder/views/shared/polling';
```
🥇 for adding this file..

pradeepdewda-okta commented on src/v2/view-builder/views/shared/polling.js:
```diff
@@ -41,15 +41,15 @@ export default {
   stopPolling() {
     if (this.polling) {
-      clearInterval(this.polling);
+      clearTimeout(this.polling);
```
good catch..

pradeepdewda-okta reviewed (approved):
@anatoliiboiarshchykov-okta Good job..!! 🛡️

haishengwu-okta commented:
Nice work @anatoliiboiarshchykov-okta ! thanks for addressing all comments!!
🚀

haishengwu-okta commented on test/testcafe/spec/DeviceChallengePollView_spec.js:
```diff
@@ -317,6 +405,32 @@ test
+  .requestHooks(loopbackPollMockLogger, loopbackPollFailedMock)('next poll should not start if previous is failed', async t => {
+    loopbackPollMockLogger.clear();
+    await setup(t);
+    await wait(10_000);
```
Are we able to wait explicit for UI changes as other tests do?

anatoliiboiarshchykov-okta replied:
Can you please link any example, to be sure that we are on the same page?

haishengwu-okta replied:
actually for this case, would `t.wait` work for you?
https://github.com/okta/okta-signin-widget/blob/master/test/testcafe/spec/EnrollAuthenticatorEmail_spec.js#L108

anatoliiboiarshchykov-okta replied:
updated

haishengwu-okta commented on test/testcafe/spec/DeviceChallengePollView_spec.js:
```diff
@@ -317,6 +405,32 @@ test
+    // Updating /poll response to take 5 sec to response.
+    // Then counting the number of calls that should be done in time interval. Default Timeout for /poll is 2 sec.
+    // Expecting to get only 2 calls(first at 2nd sec, second at 9th(5 sec response + 2 sec timeout) second).
+    await wait(10_000);
```
I may want to write test in such way
```
wait 2 second
expect poll called once
wait another 3 second
expect poll called once
wait another 10 second
expect poll called twice
```

anatoliiboiarshchykov-okta replied:
As per our conversation added a new [ticket](https://oktainc.atlassian.net/browse/OKTA-465711)
I have played about this approach a little bit and it looks like this one will be tricky(as always when you need to test something with timeline), so it is good that we decided to go with separate ticket
